### DISCUSSION TITLE
Advantage of SqlDelight over Room?

### DISCUSSION THREAD

**Discussion by kazemcodes** [UbadahJ/QReader#46] [Q&A] [answered]
Hi, I have a question, what is the advantages of sql delight over the room db, ?

also your app with 1.1b48-master is not lunching, it crashes at launch time,
   java.util.concurrent.TimeoutException
        at com.google.firebase.crashlytics.internal.common.Utils.awaitEvenIfOnMainThread(Utils.java:122)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController.handleUncaughtException(CrashlyticsController.java:232)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController$1.onUncaughtException(CrashlyticsController.java:155)
....
at com.ubadahj.qidianundergroud.MainActivity.onCreate(MainActivity.kt:59)
...
java.lang.IllegalStateException: View androidx.fragment.app.FragmentContainerView{c124d9e V.E...... ......I. 0,0-0,0 #7f0a018a app:id/nav_host_fragment} does not have a NavController set
...
 at com.ubadahj.qidianundergroud.MainActivity$onCreate$1.invokeSuspend(MainActivity.kt:69)
...
 at com.ubadahj.qidianundergroud.MainActivity.onCreate(MainActivity.kt:59)

**Comment by UbadahJ** [comment, marked as answer]
Hi!

It is more of a preference than of an advantage; Room can allow much more complex integrations with the system but SQLDelight has simple SQL files that automatically generate the relevant models for the applications. From a perspective, this is leaking of the underlying objects that are partially mitigated using Views instead of directly using Table. In comparison, Room doesn't have these flaws but requires much more boilerplate to set up. Another difference is Kotlin Multiplatform which was something I was considering at that time but clearly, that is not possible now due to use of Hilt.

To summarize, it is more of a choice decision rather than some significant advantage any of these libraries gives.

Regarding the issue, have you made a clean install? This update had a completely new DB design and there are no migrations so it requires the previous app to be uninstalled before installing the new version. If the issue persists, can you let me know further details of your device and you are using Github Release or directly building from source code?

**Reply by kazemcodes** [reply]
Im using xiaomi mi 9t android 11 and Im using the released version.

Also thank you for answering my question.

this is the error:


2022-01-27 20:14:33.479 22423-22423/com.ubadahj.qidianundergroud E/FirebaseCrashlytics: Error handling uncaught exception
    java.util.concurrent.TimeoutException
        at com.google.firebase.crashlytics.internal.common.Utils.awaitEvenIfOnMainThread(Utils.java:122)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController.handleUncaughtException(CrashlyticsController.java:232)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsController$1.onUncaughtException(CrashlyticsController.java:155)
        at com.google.firebase.crashlytics.internal.common.CrashlyticsUncaughtExceptionHandler.uncaughtException(CrashlyticsUncaughtExceptionHandler.java:54)
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
        at kotlinx.coroutines.android.AndroidExceptionPreHandler.handleException(AndroidExceptionPreHandler.kt:47)
        at kotlinx.coroutines.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:29)
        at kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:33)
        at kotlinx.coroutines.StandaloneCoroutine.handleJobException(Builders.common.kt:192)
        at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:229)
        at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:906)
        at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:863)
        at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:828)
        at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:100)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
        at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
        at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.ubadahj.qidianundergroud.MainActivity.onCreate(MainActivity.kt:59)
        at android.app.Activity.performCreate(Activity.java:8143)
        at android.app.Activity.performCreate(Activity.java:8114)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1310)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3513)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3700)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2136)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:236)
        at android.app.ActivityThread.main(ActivityThread.java:8061)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:656)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:967)
2022-01-27 20:14:33.480 22423-22423/com.ubadahj.qidianundergroud E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ubadahj.qidianundergroud, PID: 22423
    java.lang.IllegalStateException: View androidx.fragment.app.FragmentContainerView{3e962dd V.E...... ......I. 0,0-0,0 #7f0a018a app:id/nav_host_fragment} does not have a NavController set
        at androidx.navigation.Navigation.findNavController(Navigation.java:84)
        at androidx.navigation.ViewKt.findNavController(View.kt:28)
        at com.ubadahj.qidianundergroud.MainActivity$onCreate$1.invokeSuspend(MainActivity.kt:69)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:367)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
        at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
        at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at com.ubadahj.qidianundergroud.MainActivity.onCreate(MainActivity.kt:59)
        at android.app.Activity.performCreate(Activity.java:8143)
        at android.app.Activity.performCreate(Activity.java:8114)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1310)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3513)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3700)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2136)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:236)
        at android.app.ActivityThread.main(ActivityThread.java:8061)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:656)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:967)
2022-01-27 20:14:33.538 22423-22423/com.ubadahj.qidianundergroud I/Process: Sending signal. PID: 22423 SIG: 9

**Reply by UbadahJ** [reply]
Can you detail how this error is being caused? From notifications or opening the app? Also, did you clear the data of the app?

**Reply by kazemcodes** [reply]
when opening the app. yes I did clean installation.

**Reply by UbadahJ** [reply]
@kazemcodes Can you help me test the apk file and let me know if this is working for you?

<http://sendanywhe.re/GE6GUVOT>

**Reply by kazemcodes** [reply]
ok

**Reply by kazemcodes** [reply]
the past error is gone and there is a new one: 
it would be a lot better if you have a discord server.

2022-01-27 20:41:19.365 27249-27249/com.ubadahj.qidianundergroud E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ubadahj.qidianundergroud, PID: 27249
    java.lang.NullPointerException: ResultSet returned null for Book.sq:getById
        at com.squareup.sqldelight.Query.executeAsOne(Query.kt:149)
        at com.squareup.sqldelight.runtime.coroutines.FlowQuery$mapToOne$1$1.invokeSuspend(FlowExtensions.kt:61)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

**Reply by UbadahJ** [reply]
Hmm, this error should not appear on a clean install.

Regarding the Discord, since this was a pretty small project, I didn't make a discord server but give me a few minutes to do that. We can debug it on there; Until then can you uninstall and reinstall it?

**Reply by kazemcodes** [reply]
I uninstall the previous one and clean install the app and the issue is still happens.

**Reply by UbadahJ** [reply]
https://discord.gg/xVURsQnfY4

Here is an invite

**Reply by kazemcodes** [reply]
its so weird, it works perfectly when i'm using android studio to run the app the error only happens when I install the released version

**Reply by UbadahJ** [reply]
Hmm, I installed it from release and it works fine for me; The only thing I can think of is that my device is on AOSP
